File: CucumberJVM/Shared/WIMP.App/java/com/wimp/app/services/PaymentService.java
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

package com.wimp.app.services;

import com.wimp.app.data.DataRepository;
import com.wimp.app.models.Payment;
import com.wimp.app.models.Order;
import com.wimp.app.models.ServiceResult;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.stereotype.Service;

@Service
public class PaymentService {
    private final DataRepository repository;
    private final PaymentGateway paymentGateway;

    public PaymentService(DataRepository repository, PaymentGateway paymentGateway) {
        this.repository = repository;
        this.paymentGateway = paymentGateway;
    }

    public ServiceResult<Payment> authorizePaymentFor(int orderNo) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }

        try {
            Payment payment = paymentGateway.authorize(order.getCustomerEmail(), order.getPrice());
            return ServiceResult.success(payment);
        } catch (Exception ex) {
            return ServiceResult.failure(ex.getMessage());
        }
    }
}

File: Reqnroll/Shared/WIMP.App/Services/PaymentService.cs
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

using WIMP.App.Data;
using WIMP.App.Models;

namespace WIMP.App.Services;

public class PaymentService(IDataRepository repository, IPaymentGateway paymentGateway)
{
    public ServiceResult<Payment> AuthorizePaymentFor(int orderNo)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Payment>.Failure($"Order {orderNo} not found.");
        }

        try
        {
            var payment = paymentGateway.Authorize(order.CustomerEmail, order.Price);
            return ServiceResult<Payment>.Success(payment);
        }
        catch (Exception ex)
        {
            return ServiceResult<Payment>.Failure(ex.Message);
        }
    }
}

public interface IPaymentGateway
{
    public Payment Authorize(string customerEmail, decimal amount);
}

/// <summary>
/// This is a non-functioning placeholder for a real payment gateway service
/// </summary>
public class RealPaymentGateway : IPaymentGateway
{
    public Payment Authorize(string customerEmail, decimal amount)
    {
        throw new NotSupportedException("The real payment gateway cannot be used for testing");
    }
}
